### ISSUE TITLE
Thoughts on ytt as a helm templating backend?

### ISSUE THREAD

**Issue by aegershman** [k14s/ytt#40]
(Apologies, this will be a vague github issue-- about to leave for vacation)

Any thoughts on if/how ytt should/could be leveraged as a templating engine for helm charts as an "official" pluggable backend to be used for a helm chart template? Noticed there weren't any [github issues logged about it](https://github.com/helm/helm/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+ytt). And I'm under the impression the [helm `engine:` used by a chart swappable, though is `gotpl` by default (and mostly _only_ `gotpl` appears used?)](https://github.com/helm/charts/blob/master/stable/concourse/Chart.yaml#L22)

Could be completely wrong about all of this, and I have done zero research into how helm templating backends work, let alone if they're swappable at all. Just figured I'd put the question out there. Feel free to close anytime.

**Comment by nimakaviani** [comment]
@aegershman we have had conversations about having the support for ytt built into helm. 

In particular, @cppforlife and I have submitted a talk to [HELM summit Europe](https://events.linuxfoundation.org/events/helm-summit-2019/) to discuss and exercise the idea.

We also created a [bare-bone HELM plugin for ytt](https://github.com/nimakaviani/ytt-helm-plugin) a while ago but haven't really done the necessary tweaking to the core of helm to pick it up in place of the default templating engine (`gotpl`). That said, something like below should be fairly easy to achieve: 
```
helm ytt install -f template.yml -f data.yml
```

We will most likely consider the in-depth integration depending on the overall community interest. Getting the talk accepted for the HELM conference and presenting the idea is one early route that may get us there.

**Comment by cppforlife** [comment]
> That said, something like below should be fairly easy to achieve: helm ytt install -f template.yml -f data.yml

i dont really see much use unless ytt is being used as a templating engine.

i havent seen any clear docs on how to swap out templating engines., but i might not have been looking in right places.

**Comment by nimakaviani** [comment]
> i dont really see much use unless ytt is being used as a templating engine

In Helm > 3.0 it might be easier to achieve, since templating is done locally. My thinking was that with ytt as the plugin, we might be able to do an independent interpolation with ytt (without having to do surgical changes) and then reuse helm functionality for when it applies the final configs to the cluster.

**Comment by aegershman** [comment]
Hm I'm not sure if I follow @nimakaviani -- is the `helm plugin` approach referring to templating the `values.yml` which are _fed into_ a helm chart? Or about templating the helm chart itself? To clarify, I was thinking about this from the perspective of writing/templating the chart itself; e.g. when you look in a chart's `templates/` directory, the files of `templates/{deployment,service,ingress,fooblap}.yaml` would all templated using `ytt` syntax rather than [`gotmpl`](https://github.com/helm/charts/blob/master/stable/concourse/templates/web-ingress.yaml#L1-L5). Perhaps also `_helpers.tpl`.

**Comment by nimakaviani** [comment]
yup, the latter. 

The idea is to have the charts written in ytt, like [this ytt playground example](https://get-ytt.io/#example:example-k8s-helm-ish).  so, the plugin implementation (if it turns out to be practical) will delegate templating to `ytt` but will fall back to helm for `install`, `deletion` of resources, etc. 

With tiller, the plugin could work against the tiller API, something like what [helm-diff](https://github.com/databus23/helm-diff) does, but with HELM > 3.0 the plugin approach may not be as trivial. That remains to be seen as we do more investigation.

**Comment by aegershman** [comment]
Judging by the linked thread over in `helm`, I'm guessing they aren't jazzed about incorporating custom templating engines, let alone incorporating `ytt` as a templating engine directly?

It's not a huge deal but I'm just really curious-- any thoughts on where things are at right now with `ytt` and `helm`? or perhaps the current thought is to drop pursuing helm to integrate `ytt` and instead focus on other tools within `k14s`?

**Comment by cppforlife** [comment]
> Judging by the linked thread over in helm, I'm guessing they aren't jazzed about incorporating custom templating engines, let alone incorporating ytt as a templating engine directly?

yup, may be in some distant future where there is some wasm integration?

> perhaps the current thought is to drop pursuing helm to integrate ytt and instead focus on other tools within k14s?

i ve had an opinion that ytt and kapp are pretty powerful in their own right (and as combination) so it shouldnt block people from trying to adopt them.

> EDIT 12/16/2019: for anyone who stumbles on this github issue: you might want to go take a peek at https://github.com/k14s/vendir and imagine what options it could bring to the table

with ytt's library module (https://github.com/k14s/ytt/blob/develop/docs/lang-ref-ytt-library.md) and vendir you can easily manage and glue applications together. there is probably more work to do in that area but it's pretty smooth already.

**Comment by cppforlife** [comment]
this may be another interesting integration point with helm if it pans out (though it has similar difficulties as helm plugins): https://github.com/helm/helm/issues/7260
